dhis2/dhis2-core
# feat: config based data integrity checks [DHIS2-12214]

### Summary
Adds the possibility of adding configuration based integrity checks.
One such check from https://github.com/dhis2/metadata-assessment is added as a first item to test the feature.
The existing checks and non-configuration based checks such as "hard coded" in memory processing checks were adopted so that both sources of checks now only act as origin for the same generic data integrity check processing.
The existing report got adopted so that it stays "as it" externally but internally is now based on `DataIntegrityDetails` produced by each of the existing checks. 

### Compatibility
The existing checks can still be queried as report as done by the app. They can also be queried in the new API as details check. 
Referenced objects of issues are now consistently referred to by display name and UID. This was inconsistent before and some checks used as description only stating the display name.

This means in the app if one expands the details for a check this could mean the difference between (old/new)

    Commodities Amoxicillin

vs.

    Commodities Amoxicillin:ETcRJ4ozxYd

This is purely a difference in the text shown as the old report API.

### Automatic Testing
The controller based tests were extended and reorganized in more than one test class to group the growing number of tests in a meaningful way. There is now a test class for 

* tests for API returning `DataIntegritySummary`
* tests for API returning `DataIntegrityDetails`
* tests for API returning `FlattenedDataIntegrityReport`
*  tests for API returning `DataIntegrityCheck` (info about the available checks)

### Manual Testing
One should confirm that the data integrity checks in the data administration app still work: http://localhost:8080/dhis-web-data-administration/index.html#/data-integrity - ideally also setup scenarios that should come up as issues in the app.

In addition the added API needs testing, there are many scenarios:

* run all summary checks: GET http://localhost:8080/api/dataIntegrity/summary
* run individual summary checks: GET http://localhost:8080/api/dataIntegrity/summary?checks=categories-no-options
* run pattern summary checks: GET http://localhost:8080/api/dataIntegrity/summary?checks=categories*
* run single summary check: GET http://localhost:8080/api/dataIntegrity/categories-no-options/summary

and equally:

* run all details checks: GET http://localhost:8080/api/dataIntegrity/details
* run individual details checks: GET http://localhost:8080/api/dataIntegrity/details?checks=categories-no-options
* run pattern details checks: GET http://localhost:8080/api/dataIntegrity/details?checks=categories*
* run single details check: GET http://localhost:8080/api/dataIntegrity/categories-no-options/details

All of the above could be tested with other checks. The information which checks are available is returned by GET http://localhost:8080/api/dataIntegrity/ (use name)
